sample-compose/data/src/main/java/co/nimblehq/sample/compose/data/repository/AppPreferencesRepositoryImpl.kt

```diff
@@ -0,0 +1,44 @@
+package co.nimblehq.sample.compose.data.repository
+
+import android.util.Log
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.*
+import co.nimblehq.sample.compose.domain.repository.AppPreferencesRepository
+import kotlinx.coroutines.flow.*
+import java.io.IOException
+import javax.inject.Inject
+
+class AppPreferencesRepositoryImpl @Inject constructor(
+    private val appPreferencesDataStore: DataStore<Preferences>
+) : AppPreferencesRepository {
+
+    private object PreferencesKeys {
+        val FIRST_TIME_LAUNCH = booleanPreferencesKey("FIRST_TIME_LAUNCH")
+    }
+
+    override fun getFirstTimeLaunchPreferences(): Flow<Boolean> = flow {
+        val firstLaunchPreferences = appPreferencesDataStore.data
+            .catch { exception ->
+            if (exception is IOException) {
+                Log.e(
+                    AppPreferencesRepositoryImpl::class.simpleName,
+                    "Error reading preferences.",
+                    exception
+                )
+                emit(emptyPreferences())
+            } else {
+                throw exception
+            }
+        }.map { preferences ->
+            preferences[PreferencesKeys.FIRST_TIME_LAUNCH] ?: true
+        }.first()
+
+        emit(firstLaunchPreferences)
+    }
```

Comment:
@lydiasama I have an idea. Instead of `emit(emptyPreferences())` when `IOException` and then filter it to `true`. What about if we give retries(**maybe just 3 times**) 👇 , but if it's still `IOException`, then we should let it throw an error as normal. Does it make sense? 🤔 

Also, I think in this way, we don't need to `handle errors` or `emit` value here, So,  just use the `mapper` -> `flowTransform` as it handles `(error/emit)` already. What do you think? cc @nimblehq/android-chapter 
```suggestion
    override fun getFirstTimeLaunchPreferences(): Flow<Boolean> = flowTransform {
        appPreferencesDataStore.data
            .retry(retries = 3) { exception ->
                (exception is IOException).also { isIOException ->
                    if (isIOException) delay(1000)
                }
            }
            .map { preferences ->
                preferences[PreferencesKeys.FIRST_TIME_LAUNCH] ?: true
            }.first()
    }
```

Reply:
@Wadeewee Regarding retrying, I don't think we should do it. Since we don't usually do it when using the SharePreferences, so I think we should keep the same approach on the DataStore as well.

For the error mapping that you suggested, as you said `we don't need to handle errors or emit value here`, so I think we shouldn't `mapError` function from the `flowTransform`. Because even we map the error from the DataStore to that function, it will return as `else` case anyway that could waste the CPU performance. cc @nimblehq/android-chapter

Reply:
> Regarding retrying, I don't think we should do it. Since we don't usually do it when using the SharePreferences

@lydiasama, In `SharePreferences` we don't need a handler `IOExeption` when read/write. That's why I said It's weird that the `code lab` suggested handling `IOExeption` when using `DataStore`, and because we have handled `IOExeption`, that's why I suggest retrying to avoid emitting `empty` value. 🙏 


> so I think we should keep the same approach on the DataStore as well.

Regarding this ☝️ If we use the same approach, May I know how we handle this case:

- When entering the app but `not for the first time`, That's mean we have set `false` for  `updateFirstTimeLaunch` once we enter it for the first time.
- And then we get `IOExeption` (**The current value should be false**), but whenever `IOExeption`, we will emit `emptyPreferences()` and then map the value to `true`. Is that mean we will show `toast` and `update value` again? 🤔 

Please correct me if I missed something. cc @Tuubz

Reply:
@Wadeewee Whenever you call a function, it will use the CPU every time. And for this case, when you go into the `mapError` function, it will return the same exception you pass. So it's useless to do that. 😅

Reply:
> I don't think we need to repeat the fetching 3 times here

Oh, @kaungkhantsoe, I don't suppose to retry 3 times as well,  just give an example of using `retry` to handling `IOExeption`

Reply:
@Wadeewee I think @kaungkhantsoe  means when it returns error `IOException`, it means there was an issue about the device. So even if we retry to connect the device again, it will return the same `IOException` forever. Hence, we don't need to keep retrying it.

Reply:
@lydiasama, TBH, I don't think we need to handle `IOException` for the current approach. It shouldn't take more time to get(**read**) the value that causes it `IOException`. It would be great if we could reproduce it.  but if we decide to emit `empty` when `IOException`, What do you think about this case? Can it be happen?

> - When entering the app but not for the first time, That's mean we have set false for updateFirstTimeLaunch once we enter it for the first time.
> - And then we get IOExeption (The current value should be false), but whenever IOExeption, we will emit emptyPreferences() and then map the value to true. Is that mean we will show toast and update value again? 🤔

cc @Tuubz

Reply:
@Wadeewee 
> And then we get IOExeption (The current value should be false), but whenever IOExeption, we will emit emptyPreferences() and then map the value to true. Is that mean we will show toast and update value again? 

About this case, I think if it gets IOException, then returning the default preferences > return `true` makes sense to me. Since the IOException handling is recommended by Google's codelab, so I think we should handle it too.

Reply:
@Wadeewee According to [this issue tracker](https://issuetracker.google.com/issues/173522155), not using a single instance of the datastore can cause `IOException`. cc @lydiasama

Reply:
@kaungkhantsoe I think I implemented the single instance of the datastore already. 🤔 You can see in the module, I used `@singleton`. Or did I miss something?

Reply:
@lydiasama I am just answering what could cause `IOException` with `DataStore`. 😬 You did it right 👍

Reply:
> About this case, I think if it gets IOException, then returning the default preferences > return true makes sense to me. 

@lydiasama, I just want to clarify this ☝️, So are you saying, It makes sense to `show a toast` with the message `firstInstall` again, even if it's not a first install, right? 🙏

Reply:
@Wadeewee It's correct. That's my intention to handle it that way. Since this is just an example, we don't have any requirements for handling it. So let's keep it simple pls. 🙏🏻